File: src/Organon/Syl/Proof.hs
module Organon.Syl.Proof
  ( reduce,
  )
where

import Organon.Syl.Proposition (contradictory)
import Organon.Syl.Types

-- | Construct a reduction proof for a valid syllogism of the given mood.
-- Returns the sequence of proof steps that reduce it to a Figure I syllogism.
--
-- Precondition: the syllogism must actually be valid in the given mood.
-- If called on an invalid syllogism, the proof steps will be nonsensical.
reduce :: Mood -> Syllogism -> [ProofStep]
-- Figure I: perfect syllogisms, accepted as axioms.
reduce Barbara _ = [Axiom Barbara]
reduce Celarent _ = [Axiom Celarent]
reduce Darii _ = [Axiom Darii]
reduce Ferio _ = [Axiom Ferio]
-- Figure II

-- Cesare (EAE-2) → Celarent: simple convert major.
--   E(P,M), A(S,M) → E(S,P)
--   Convert E(P,M) → E(M,P), then Celarent.
reduce Cesare (Syllogism maj _ _) =
  let maj' = swapTerms maj
   in [SimpleConversion maj maj', Axiom Celarent]
-- Camestres (AEE-2) → Celarent: simple convert minor, mutate, Celarent,
--   simple convert conclusion.
--   A(P,M), E(S,M) → E(S,P)
reduce Camestres (Syllogism maj min_ concl) =
  let min' = swapTerms min_
      derivedConcl = Proposition E (subject maj) (predicate min')
   in [ SimpleConversion min_ min',
        MutatePremises,
        Axiom Celarent,
        SimpleConversion derivedConcl concl
      ]
-- Festino (EIO-2) → Ferio: simple convert major.
--   E(P,M), I(S,M) → O(S,P)
--   Convert E(P,M) → E(M,P), then Ferio.
reduce Festino (Syllogism maj _ _) =
  let maj' = swapTerms maj
   in [SimpleConversion maj maj', Axiom Ferio]
-- Baroco (AOO-2) → Barbara: reductio ad impossibile.
--   A(P,M), O(S,M) → O(S,P)
--   Assume A(S,P) [contradictory of O(S,P)].
--   Barbara: A(P,M), A(S,P) → A(S,M), which contradicts O(S,M).
reduce Baroco (Syllogism maj min_ concl) =
  let assumed = contradictory concl
      derived = Proposition A (subject min_) (predicate maj)
      reductioSyl = Syllogism maj assumed derived
   in [ReductioAdImpossibile Barbara assumed reductioSyl]
-- Figure III

-- Darapti (AAI-3) → Darii: convert minor per accidens.
--   A(M,P), A(M,S) → I(S,P)
--   Convert A(M,S) → I(S,M), then Darii.
reduce Darapti (Syllogism _ min_ _) =
  let min' = Proposition I (predicate min_) (subject min_)
   in [ConversionPerAccidens min_ min', Axiom Darii]
-- Disamis (IAI-3) → Darii: simple convert major, mutate, Darii,
--   simple convert conclusion.
--   I(M,P), A(M,S) → I(S,P)
reduce Disamis (Syllogism maj min_ concl) =
  let maj' = swapTerms maj
      derivedConcl = Proposition I (subject min_) (predicate maj')
   in [ SimpleConversion maj maj',
        MutatePremises,
        Axiom Darii,
        SimpleConversion derivedConcl concl
      ]
-- Datisi (AII-3) → Darii: simple convert minor.
--   A(M,P), I(M,S) → I(S,P)
--   Convert I(M,S) → I(S,M), then Darii.
reduce Datisi (Syllogism _ min_ _) =
  let min' = swapTerms min_
   in [SimpleConversion min_ min', Axiom Darii]
-- Felapton (EAO-3) → Ferio: convert minor per accidens.
--   E(M,P), A(M,S) → O(S,P)
--   Convert A(M,S) → I(S,M), then Ferio.
reduce Felapton (Syllogism _ min_ _) =
  let min' = Proposition I (predicate min_) (subject min_)
   in [ConversionPerAccidens min_ min', Axiom Ferio]
-- Bocardo (OAO-3) → Barbara: reductio ad impossibile.
--   O(M,P), A(M,S) → O(S,P)
--   Assume A(S,P) [contradictory of O(S,P)].
--   Barbara: A(S,P), A(M,S) → A(M,P), which contradicts O(M,P).
reduce Bocardo (Syllogism maj min_ concl) =
  let assumed = contradictory concl
      derived = Proposition A (subject maj) (predicate maj)
      reductioSyl = Syllogism assumed min_ derived
   in [ReductioAdImpossibile Barbara assumed reductioSyl]
-- Ferison (EIO-3) → Ferio: simple convert minor.
--   E(M,P), I(M,S) → O(S,P)
--   Convert I(M,S) → I(S,M), then Ferio.
reduce Ferison (Syllogism _ min_ _) =
  let min' = swapTerms min_
   in [SimpleConversion min_ min', Axiom Ferio]
-- Figure IV

-- Bramantip (AAI-4) → Barbara: mutate, Barbara, convert conclusion per accidens.
--   A(P,M), A(M,S) → I(S,P)
reduce Bramantip (Syllogism maj min_ concl) =
  let derivedConcl = Proposition A (subject maj) (predicate min_)
   in [ MutatePremises,
        Axiom Barbara,
        ConversionPerAccidens derivedConcl concl
      ]
-- Camenes (AEE-4) → Celarent: mutate, Celarent, simple convert conclusion.
--   A(P,M), E(M,S) → E(S,P)
reduce Camenes (Syllogism maj min_ concl) =
  let derivedConcl = Proposition E (subject maj) (predicate min_)
   in [ MutatePremises,
        Axiom Celarent,
        SimpleConversion derivedConcl concl
      ]
-- Dimaris (IAI-4) → Darii: mutate, Darii, simple convert conclusion.
--   I(P,M), A(M,S) → I(S,P)
reduce Dimaris (Syllogism maj min_ concl) =
  let derivedConcl = Proposition I (subject maj) (predicate min_)
   in [ MutatePremises,
        Axiom Darii,
        SimpleConversion derivedConcl concl
      ]
-- Fesapo (EAO-4) → Ferio: simple convert major, convert minor per accidens.
--   E(P,M), A(M,S) → O(S,P)
reduce Fesapo (Syllogism maj min_ _) =
  let maj' = swapTerms maj
      min' = Proposition I (predicate min_) (subject min_)
   in [ SimpleConversion maj maj',
        ConversionPerAccidens min_ min',
        Axiom Ferio
      ]
-- Fresison (EIO-4) → Ferio: simple convert major, simple convert minor.
--   E(P,M), I(M,S) → O(S,P)
reduce Fresison (Syllogism maj min_ _) =
  let maj' = swapTerms maj
      min' = swapTerms min_
   in [ SimpleConversion maj maj',
        SimpleConversion min_ min',
        Axiom Ferio
      ]
-- Subaltern moods: reduce the parent mood, then subalternate the conclusion.

reduce Barbari syl =
  reduceSubaltern Barbara A syl
reduce Celaront syl =
  reduceSubaltern Celarent E syl
reduce Cesaro syl =
  reduceSubaltern Cesare E syl
reduce Camestrop syl =
  reduceSubaltern Camestres E syl
reduce Calemos syl =
  reduceSubaltern Camenes E syl

-- Helpers

-- | Swap subject and predicate of a proposition, keeping the type.
swapTerms :: Proposition -> Proposition
swapTerms (Proposition t s p) = Proposition t p s

-- | Reduce a subaltern mood by reducing the parent mood with a stronger
-- conclusion, then subalternating.
reduceSubaltern :: Mood -> PropType -> Syllogism -> [ProofStep]
reduceSubaltern parent strongType (Syllogism maj min_ concl) =
  let strongConcl = Proposition strongType (subject concl) (predicate concl)
      parentSyl = Syllogism maj min_ strongConcl
   in reduce parent parentSyl ++ [Subalternation strongConcl concl]

File: web/src/core/proof.ts
// Reduction proofs — ported from Organon.Syl.Proof

import { contradictory } from "./proposition";
import { Mood, PropType, Proposition, ProofStep, Syllogism } from "./types";

function swapTerms(p: Proposition): Proposition {
  return { propType: p.propType, subject: p.predicate, predicate: p.subject };
}

function reduceSubaltern(
  parent: Mood,
  strongType: PropType,
  syl: Syllogism,
): ProofStep[] {
  const strongConcl: Proposition = {
    propType: strongType,
    subject: syl.conclusion.subject,
    predicate: syl.conclusion.predicate,
  };
  const parentSyl: Syllogism = {
    major: syl.major,
    minor: syl.minor,
    conclusion: strongConcl,
  };
  return [
    ...reduce(parent, parentSyl),
    { tag: "Subalternation", from: strongConcl, to: syl.conclusion },
  ];
}

export function reduce(mood: Mood, syl: Syllogism): ProofStep[] {
  const { major: maj, minor: min, conclusion: concl } = syl;

  switch (mood) {
    // Figure I: perfect syllogisms
    case Mood.Barbara:
      return [{ tag: "Axiom", mood: Mood.Barbara }];
    case Mood.Celarent:
      return [{ tag: "Axiom", mood: Mood.Celarent }];
    case Mood.Darii:
      return [{ tag: "Axiom", mood: Mood.Darii }];
    case Mood.Ferio:
      return [{ tag: "Axiom", mood: Mood.Ferio }];

    // Figure II
    case Mood.Cesare: {
      const maj2 = swapTerms(maj);
      return [
        { tag: "SimpleConversion", from: maj, to: maj2 },
        { tag: "Axiom", mood: Mood.Celarent },
      ];
    }
    case Mood.Camestres: {
      const min2 = swapTerms(min);
      const derivedConcl: Proposition = {
        propType: PropType.E,
        subject: maj.subject,
        predicate: min2.predicate,
      };
      return [
        { tag: "SimpleConversion", from: min, to: min2 },
        { tag: "MutatePremises" },
        { tag: "Axiom", mood: Mood.Celarent },
        { tag: "SimpleConversion", from: derivedConcl, to: concl },
      ];
    }
    case Mood.Festino: {
      const maj2 = swapTerms(maj);
      return [
        { tag: "SimpleConversion", from: maj, to: maj2 },
        { tag: "Axiom", mood: Mood.Ferio },
      ];
    }
    case Mood.Baroco: {
      const assumed = contradictory(concl);
      const derived: Proposition = {
        propType: PropType.A,
        subject: min.subject,
        predicate: maj.predicate,
      };
      const reductioSyl: Syllogism = {
        major: maj,
        minor: assumed,
        conclusion: derived,
      };
      return [
        {
          tag: "ReductioAdImpossibile",
          targetMood: Mood.Barbara,
          assumed,
          syllogism: reductioSyl,
        },
      ];
    }

    // Figure III
    case Mood.Darapti: {
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return [
        { tag: "ConversionPerAccidens", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Darii },
      ];
    }
    case Mood.Disamis: {
      const maj2 = swapTerms(maj);
      const derivedConcl: Proposition = {
        propType: PropType.I,
        subject: min.subject,
        predicate: maj2.predicate,
      };
      return [
        { tag: "SimpleConversion", from: maj, to: maj2 },
        { tag: "MutatePremises" },
        { tag: "Axiom", mood: Mood.Darii },
        { tag: "SimpleConversion", from: derivedConcl, to: concl },
      ];
    }
    case Mood.Datisi: {
      const min2 = swapTerms(min);
      return [
        { tag: "SimpleConversion", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Darii },
      ];
    }
    case Mood.Felapton: {
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return [
        { tag: "ConversionPerAccidens", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Ferio },
      ];
    }
    case Mood.Bocardo: {
      const assumed = contradictory(concl);
      const derived: Proposition = {
        propType: PropType.A,
        subject: maj.subject,
        predicate: maj.predicate,
      };
      const reductioSyl: Syllogism = {
        major: assumed,
        minor: min,
        conclusion: derived,
      };
      return [
        {
          tag: "ReductioAdImpossibile",
          targetMood: Mood.Barbara,
          assumed,
          syllogism: reductioSyl,
        },
      ];
    }
    case Mood.Ferison: {
      const min2 = swapTerms(min);
      return [
        { tag: "SimpleConversion", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Ferio },
      ];
    }

    // Figure IV
    case Mood.Bramantip: {
      const derivedConcl: Proposition = {
        propType: PropType.A,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return [
        { tag: "MutatePremises" },
        { tag: "Axiom", mood: Mood.Barbara },
        { tag: "ConversionPerAccidens", from: derivedConcl, to: concl },
      ];
    }
    case Mood.Camenes: {
      const derivedConcl: Proposition = {
        propType: PropType.E,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return [
        { tag: "MutatePremises" },
        { tag: "Axiom", mood: Mood.Celarent },
        { tag: "SimpleConversion", from: derivedConcl, to: concl },
      ];
    }
    case Mood.Dimaris: {
      const derivedConcl: Proposition = {
        propType: PropType.I,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return [
        { tag: "MutatePremises" },
        { tag: "Axiom", mood: Mood.Darii },
        { tag: "SimpleConversion", from: derivedConcl, to: concl },
      ];
    }
    case Mood.Fesapo: {
      const maj2 = swapTerms(maj);
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return [
        { tag: "SimpleConversion", from: maj, to: maj2 },
        { tag: "ConversionPerAccidens", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Ferio },
      ];
    }
    case Mood.Fresison: {
      const maj2 = swapTerms(maj);
      const min2 = swapTerms(min);
      return [
        { tag: "SimpleConversion", from: maj, to: maj2 },
        { tag: "SimpleConversion", from: min, to: min2 },
        { tag: "Axiom", mood: Mood.Ferio },
      ];
    }

    // Subaltern moods
    case Mood.Barbari:
      return reduceSubaltern(Mood.Barbara, PropType.A, syl);
    case Mood.Celaront:
      return reduceSubaltern(Mood.Celarent, PropType.E, syl);
    case Mood.Cesaro:
      return reduceSubaltern(Mood.Cesare, PropType.E, syl);
    case Mood.Camestrop:
      return reduceSubaltern(Mood.Camestres, PropType.E, syl);
    case Mood.Calemos:
      return reduceSubaltern(Mood.Camenes, PropType.E, syl);
  }
}

function reducedSubaltern(
  parent: Mood,
  strongType: PropType,
  syl: Syllogism,
): Syllogism | null {
  const strongConcl: Proposition = {
    propType: strongType,
    subject: syl.conclusion.subject,
    predicate: syl.conclusion.predicate,
  };
  const parentSyl: Syllogism = {
    major: syl.major,
    minor: syl.minor,
    conclusion: strongConcl,
  };
  const r = reducedSyllogism(parent, parentSyl);
  return r !== null ? r : parentSyl;
}

/** Compute the Figure 1 syllogism that a mood reduces to.
 *  Returns null for Figure 1 moods (already there) and reductio
 *  moods (Baroco, Bocardo) which have no direct Figure 1 equivalent. */
export function reducedSyllogism(mood: Mood, syl: Syllogism): Syllogism | null {
  const { major: maj, minor: min, conclusion: concl } = syl;

  switch (mood) {
    // Figure I: already there.
    case Mood.Barbara:
    case Mood.Celarent:
    case Mood.Darii:
    case Mood.Ferio:
      return null;

    // Figure II
    case Mood.Cesare:
      return { major: swapTerms(maj), minor: min, conclusion: concl };
    case Mood.Camestres: {
      const min2 = swapTerms(min);
      const derivedConcl: Proposition = {
        propType: PropType.E,
        subject: maj.subject,
        predicate: min2.predicate,
      };
      return { major: min2, minor: maj, conclusion: derivedConcl };
    }
    case Mood.Festino:
      return { major: swapTerms(maj), minor: min, conclusion: concl };
    case Mood.Baroco:
      return null;

    // Figure III
    case Mood.Darapti: {
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return { major: maj, minor: min2, conclusion: concl };
    }
    case Mood.Disamis: {
      const maj2 = swapTerms(maj);
      const derivedConcl: Proposition = {
        propType: PropType.I,
        subject: min.subject,
        predicate: maj2.predicate,
      };
      return { major: min, minor: maj2, conclusion: derivedConcl };
    }
    case Mood.Datisi:
      return { major: maj, minor: swapTerms(min), conclusion: concl };
    case Mood.Felapton: {
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return { major: maj, minor: min2, conclusion: concl };
    }
    case Mood.Bocardo:
      return null;
    case Mood.Ferison:
      return { major: maj, minor: swapTerms(min), conclusion: concl };

    // Figure IV
    case Mood.Bramantip: {
      const derivedConcl: Proposition = {
        propType: PropType.A,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return { major: min, minor: maj, conclusion: derivedConcl };
    }
    case Mood.Camenes: {
      const derivedConcl: Proposition = {
        propType: PropType.E,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return { major: min, minor: maj, conclusion: derivedConcl };
    }
    case Mood.Dimaris: {
      const derivedConcl: Proposition = {
        propType: PropType.I,
        subject: maj.subject,
        predicate: min.predicate,
      };
      return { major: min, minor: maj, conclusion: derivedConcl };
    }
    case Mood.Fesapo: {
      const maj2 = swapTerms(maj);
      const min2: Proposition = {
        propType: PropType.I,
        subject: min.predicate,
        predicate: min.subject,
      };
      return { major: maj2, minor: min2, conclusion: concl };
    }
    case Mood.Fresison:
      return {
        major: swapTerms(maj),
        minor: swapTerms(min),
        conclusion: concl,
      };

    // Subaltern moods
    case Mood.Barbari:
      return reducedSubaltern(Mood.Barbara, PropType.A, syl);
    case Mood.Celaront:
      return reducedSubaltern(Mood.Celarent, PropType.E, syl);
    case Mood.Cesaro:
      return reducedSubaltern(Mood.Cesare, PropType.E, syl);
    case Mood.Camestrop:
      return reducedSubaltern(Mood.Camestres, PropType.E, syl);
    case Mood.Calemos:
      return reducedSubaltern(Mood.Camenes, PropType.E, syl);
  }
}
